File: LeetCode/0407-Trapping-Rain-Water-II.cpp
// 407. Trapping Rain Water II

class Solution {
public:
    int trapRainWater(vector<vector<int>>& heightMap) {
        int n = heightMap.size(), m = heightMap[0].size();

        if(n <= 2 || m <= 2) return 0;
        
        priority_queue<vector<int>, vector<vector<int>>, greater<vector<int>>> pq;
        vector<vector<bool>> vis(n, vector<bool>(m, false));

        for(int i = 0; i < n; i++) {
            pq.push({heightMap[i][0], i, 0});
            vis[i][0] = true;
            pq.push({heightMap[i][m - 1], i, m - 1});
            vis[i][m - 1] = true;
        }
        
        for(int j = 0; j < m; j++) {
            pq.push({heightMap[0][j], 0, j});
            vis[0][j] = true;
            pq.push({heightMap[n - 1][j], n - 1, j});
            vis[n - 1][j] = true;
        }

        vector<int> dx = {+1, 0, -1, 0};
        vector<int> dy = {0, +1, 0, -1};

        int trapped_water = 0;
        while(!pq.empty()) {
            auto curr = pq.top();
            pq.pop();

            int h = curr[0], x = curr[1], y = curr[2];
            for(int k = 0; k < 4; k++) {
                int nx = x + dx[k], ny = y + dy[k];
                if(nx >= 0 && nx < n && ny >= 0 && ny < m && !vis[nx][ny]) {
                    vis[nx][ny] = true;
                    trapped_water += max(0, h - heightMap[nx][ny]);
                    pq.push({max(h, heightMap[nx][ny]), nx, ny});
                }
            }
        }

        return trapped_water;
    }
};

File: LeetCode/0407-Trapping-Rain-Water-II.java
// 407. Trapping Rain Water II

class Cell {
    int row, col, height;
    public Cell(int row, int col, int height) {
        this.row = row;
        this.col = col;
        this.height = height;
    }
}

class Solution {
    int[] dx = {-1, 0, +1, 0};
    int[] dy = {0, +1, 0, -1};

    Comparator<Cell> cmp = (a, b) -> {
        return a.height - b.height;
    };

    public int trapRainWater(int[][] heightMap) {
        int n = heightMap.length, m = heightMap[0].length;
        if (n == 1 || m == 1) return 0;

        boolean[][] vis = new boolean[n][m];
        PriorityQueue<Cell> pq = new PriorityQueue<>(cmp);
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (i == 0 || i == n - 1 || j == 0 || j == m - 1) {
                    pq.offer(new Cell(i, j, heightMap[i][j]));
                    vis[i][j] = true;
                }
            }
        }

        int water = 0;
        while (!pq.isEmpty()) {
            Cell curr = pq.poll();
            int x = curr.row, y = curr.col, h = curr.height;
            for (int dir = 0; dir < 4; dir++) {
                int nx = x + dx[dir];
                int ny = y + dy[dir];
                if (nx >= 0 && nx < n && ny >= 0 && ny < m && !vis[nx][ny]) {
                    vis[nx][ny] = true;
                    if (h > heightMap[nx][ny]) water += h - heightMap[nx][ny];
                    pq.offer(new Cell(nx, ny, Math.max(h, heightMap[nx][ny])));
                }
            }
        }

        return water;
    }
}
